### PR TITLE
Twitch studio importer

### PR DESCRIPTION
**Pull request by andycreeth**

### CODE REVIEW

File: app/services/ts-importer.ts [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    if (item) {
+      const targetWidth =
+        (1 - layer.position.right - layer.position.left) *
+        this.videoService.baseResolutions.horizontal.baseWidth;
+      const targetHeight =
+        (1 - layer.position.bottom - layer.position.top) *
+        this.videoService.baseResolutions.horizontal.baseHeight;
+
+      const sourceWidth = widthOverride ?? item.getSource().width;
+      const sourceHeight = widthOverride ?? item.getSource().height;
```
[the remark this answers is not in the document]
Comment (andycreeth) [reply app/services/ts-importer.ts:360]: woahhhh good catch.  I was seeing some weird crop behavior I didn't understand, I bet this is why!  thank you!

File: app/services/ts-importer.ts [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+    let item: SceneItem;
+    let noCrop = false;
+    let widthOverride: number;
+    let heightOverride: number;
+
+    if (node.pluginId === 'windowsVideoCapture') {
+      let webcamSource = this.sourcesService.views.getSourcesByType('dshow_input')[0];
+
+      if (!webcamSource) {
+        webcamSource = this.sourcesService.createSource(layer.name, 'dshow_input');
```
Comment (blackxored) [inline app/services/ts-importer.ts:243]: Don't we have a specialized method for creating these that's supposedly abstracting away the differences between Windows and MacOS? I might be thinking of a different source type, but I remember seeing this somewhere.

**Review by gettinToasty — APPROVED** [review]

File: app/services/ts-importer.ts (hunk 2 above)
Comment (andycreeth) [reply app/services/ts-importer.ts:247]: It's not wrapped in a method, but we do have that code somewhere.  This importer is windows only though.

File: app/services/ts-importer.ts [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+      const y = this.videoService.baseResolutions.horizontal.baseHeight * layer.position.top;
+
+      item.setTransform({ scale: { x: scale, y: scale }, position: { x, y }, crop });
+    }
+  }
+
+  // Waits for a source to become a non-zero size, with a timeout
+  waitForNonzeroSize(source: Source) {
+    return new Promise<void>(resolve => {
+      const sub = this.sourcesService.sourceUpdated.subscribe(s => {
```
[the remark this answers is not in the document]
Comment (andycreeth) [reply app/services/ts-importer.ts:400]: Yeah RxJS has a lot of fanciness that I'm not familiar with and should learn at some point.

File: app/services/ts-importer.ts [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        this.setupVideo(config);
+        this.importScenes(config);
+
+        return this.scenesService.views.scenes.length !== 0;
+      },
+    });
+  }
+
+  setupVideo(config: ITSConfig) {
+    this.settingsService.setSettingValue(
```
[the remark this answers is not in the document]
Comment (andycreeth) [reply app/services/ts-importer.ts:200]: Ohh interesting I didn't know that.

**Pull request merged**